databroker/src/grpc/kuksa_val_v2/val.rs

```diff
@@ -84,11 +86,88 @@ impl proto::val_server::Val for broker::DataBroker {
         Err(tonic::Status::new(Code::Unimplemented, "Unimplemented"))
     }
 
+    /// List metadata of signals matching the wildcard branch request.
+    ///
+    /// # Arguments
+    ///
+    ///      ```
+    ///     `request`:
+    ///      ListMetadataRequest {
+    ///          root: String
+    ///          filter: String
+    ///      }
+    ///
+    /// # Response
+    ///      `response`:
+    ///      ListMetadataResponse {
+    ///          metadata: Vec<Metadata>
+    ///      }
+    ///      ```
+    ///
+    /// # Errors
+    ///
+    /// Returns (GRPC error code):
+    /// NOT_FOUND if the specified root branch does not exist
+    /// INVALID_ARGUMENT if the request pattern is invalid
+    ///
+    /// # Examples
+    /// For details, please refer to
+    /// [Wildcard Matching](https://github.com/eclipse-kuksa/kuksa-databroker/blob/main/doc/wildcard_matching.md#examples)
     async fn list_metadata(
         &self,
-        _request: tonic::Request<proto::ListMetadataRequest>,
+        request: tonic::Request<proto::ListMetadataRequest>,
     ) -> Result<tonic::Response<proto::ListMetadataResponse>, tonic::Status> {
-        Err(tonic::Status::new(Code::Unimplemented, "Unimplemented"))
+        debug!(?request);
+        let permissions = match request.extensions().get::<Permissions>() {
+            Some(permissions) => {
+                debug!(?permissions);
+                permissions.clone()
+            }
+            None => return Err(tonic::Status::unauthenticated("Unauthenticated")),
+        };
+        let broker = self.authorized_access(&permissions);
+
+        let metadata_request = request.into_inner();
+
+        match Matcher::new(&metadata_request.root) {
+            Ok(matcher) => {
+                let mut metadata_response = Vec::new();
+                broker
+                    .for_each_entry(|entry| {
+                        let entry_metadata = &entry.metadata();
+                        if matcher.is_match(&entry_metadata.glob_path) {
+                            metadata_response.push(Metadata {
+                                id: entry_metadata.id,
+                                data_type: proto::DataType::from(entry_metadata.data_type.clone())
+                                    as i32,
+                                entry_type: proto::EntryType::from(
+                                    entry_metadata.entry_type.clone(),
+                                ) as i32,
+                                description: Some(entry_metadata.description.clone()),
+                                comment: None,
+                                deprecation: None,
+                                unit: entry_metadata.unit.clone(),
+                                value_restriction: None,
```

Comment:
why are some things None shouldn't we extend databroker to these attributes? e.g. add them to DataEntry?

Reply:
Maybe we should bring this topic to API discussions. It seems to be also present in V1 but never used